pkg/ui/workspaces/cluster-ui/src/store/jobDetails/job.selectors.ts

```diff
@@ -9,9 +9,53 @@
 // licenses/APL.txt.
 
 import { createSelector } from "reselect";
-import { adminUISelector } from "../utils/selectors";
+import { adminUISelector } from "src/store/utils/selectors";
+import { selectID } from "src/selectors";
+import { JobResponse } from "src/api/jobsApi";
 
-export const selectJobState = createSelector(
-  adminUISelector,
-  adminUiState => adminUiState.job,
+const selectJobState = createSelector(adminUISelector, state => {
+  const jobState = state?.job?.cachedData;
+  const emptyJobCache = !jobState || Object.keys(jobState).length === 0;
+  if (emptyJobCache) {
+    return null;
+  }
+  return jobState;
+});
+
+export const selectJob = createSelector(
+  [adminUISelector, selectJobState, selectID],
+  (adminUIState, jobState, jobID) => {
+    const jobsCache = adminUIState?.jobs?.data;
+    let job: JobResponse;
+    if (!jobID || (!jobsCache && !jobState)) {
+      return null;
+    } else if (jobsCache) {
+      job = Object(jobsCache.jobs.find(job => job.id.toString() === jobID));
+    } else if (jobState) {
+      job = jobState[jobID]?.data;
```

Comment:
I know this has merged already, but my previous comment was envisioning something more like:
- check the job details cache first (if an entry exists there, we should just use it)
- If no entry in cache for this job id, check the overview cache and attempt to find the job 
We might even be able to get away with setting jobState[jobID] = foundJobsFromOverview here, but I don't know if that's very appropriate/idiomatic. Also, why do we need the Object constructor in line 33?

Reply:
> check the job details cache first (if an entry exists there, we should just use it)

In retrospect, this makes a lot more sense, and I don't know why I didn't do it in the first place :grin: . https://github.com/cockroachdb/cockroach/issues/96246

> We might even be able to get away with setting jobState[jobID] = foundJobsFromOverview here, but I don't know if that's very appropriate/idiomatic.

I did spend a little while thinking about where/how to do this properly, but IDK if it's necessary if we are just going to select from the Jobs cache if there isn't anything in the Job details cache, and then when we poll every 10 (?) seconds, the results just overwrite whatever is in the Job details cache. So writing jobState[jobID] = foundJobsFromOverview wouldn't save us a request (unless I'm missing something). Also, looking at the redux docs, it seems like selectors are very much only meant for deriving state, and not for setting it, but :man_shrugging: .

> Also, why do we need the Object constructor in line 33?

The array of jobs returned in a `JobsResponse` is typed differently (`cockroach.server.serverpb.IJobResponse[]` vs. `JobResponse`). `JobResponse` implements `IJobResponse`. Without the Object constructor on what's returned in `jobsCache.jobs.find(job => job.id.toString() === jobID)`, the build fails on a type error.